src/python/pants/backend/python/interpreter_cache.py

```diff
@@ -74,29 +74,34 @@ def _cache_dir(self):
 
   def select_interpreter_for_targets(self, targets):
     """Pick an interpreter compatible with all the specified targets."""
-    tgts_with_compatibilities = []
+    tgts_by_compatibilities = {}
     filters = set()
     for target in targets:
-      if isinstance(target, PythonTarget) and target.compatibility:
-        tgts_with_compatibilities.append(target)
-        filters.update(target.compatibility)
+      if isinstance(target, PythonTarget):
+        c = self._python_setup.compatibility_or_constraintstar(target)
+        targets = tgts_by_compatibilities.get(c)
```

Comment:
Could you not streamline all this with a `defaultdict(list)`? Then `tgts_by_compatibilities[c]` will create, assign and return an empty list if none exists at that key.